File: benchmark/go/fib.go
// fib.go - Fibonacci Benchmark for Go
// Tests function call overhead (recursive) and loop performance (iterative)
package main

import (
	"fmt"
	"time"
)

// Recursive Fibonacci - tests function call overhead
func fibRecursive(n int) int64 {
	if n <= 1 {
		return int64(n)
	}
	return fibRecursive(n-1) + fibRecursive(n-2)
}

// Iterative Fibonacci - tests loop performance
func fibIterative(n int) int64 {
	if n <= 1 {
		return int64(n)
	}
	var a, b int64 = 0, 1
	for i := 2; i <= n; i++ {
		a, b = b, a+b
	}
	return b
}

func main() {
	// Recursive fib(35)
	start := time.Now()
	result := fibRecursive(35)
	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("Recursive fib(35) = %d\n", result)
	fmt.Printf("Recursive time: %dms\n", elapsed)

	// Iterative fib(50)
	start = time.Now()
	result = fibIterative(50)
	elapsed = time.Since(start).Milliseconds()
	fmt.Printf("Iterative fib(50) = %d\n", result)
	fmt.Printf("Iterative time: %dms\n", elapsed)
}

File: benchmark/rust/fib.rs
// fib.rs - Fibonacci Benchmark for Rust
// Tests function call overhead (recursive) and loop performance (iterative)

use std::time::Instant;

// Recursive Fibonacci - tests function call overhead
fn fib_recursive(n: i32) -> i64 {
    if n <= 1 {
        return n as i64;
    }
    fib_recursive(n - 1) + fib_recursive(n - 2)
}

// Iterative Fibonacci - tests loop performance
fn fib_iterative(n: i32) -> i64 {
    if n <= 1 {
        return n as i64;
    }
    let mut a: i64 = 0;
    let mut b: i64 = 1;
    for _ in 2..=n {
        let temp = a + b;
        a = b;
        b = temp;
    }
    b
}

fn main() {
    // Recursive fib(35)
    let start = Instant::now();
    let result = fib_recursive(35);
    let elapsed = start.elapsed().as_millis();
    println!("Recursive fib(35) = {}", result);
    println!("Recursive time: {}ms", elapsed);

    // Iterative fib(50)
    let start = Instant::now();
    let result = fib_iterative(50);
    let elapsed = start.elapsed().as_millis();
    println!("Iterative fib(50) = {}", result);
    println!("Iterative time: {}ms", elapsed);
}
